### PR TITLE
[lang] Add a method to get all SNodes under a root

### PR DESCRIPTION
**Pull request by k-ye**
Related issue = #3018

Adding this helper to make it easier for us to implement multi-root support

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -53,7 +53,7 @@ include(cmake/TaichiCXXFlags.cmake)
 include(cmake/TaichiCore.cmake)
 include(cmake/TaichiMain.cmake)
 
-option(TI_BUILD_TESTS "Build the CPP tests" OFF)
+option(TI_BUILD_TESTS "Build the CPP tests" ON)
```
Comment (k-ye) [inline CMakeLists.txt:56]:
```suggestion
option(TI_BUILD_TESTS "Build the CPP tests" OFF)
```

File: taichi/struct/snode_tree.cpp [hunk 2]
```diff
@@ -2,6 +2,18 @@
 
 namespace taichi {
 namespace lang {
+namespace {
+
+void get_snodes_to_root_id_impl(const SNode &node,
+                                const int root_id,
+                                std::unordered_map<int, int> *map) {
```
Comment (strongoier) [inline taichi/struct/snode_tree.cpp:9]: Is there a reason we use the `*`-way instead of `&`-way to pass the map?

**Review by strongoier — APPROVED** [review]
LGTM!

**Review by ailzhang — APPROVED** [review]
LGTM!

File: taichi/struct/snode_tree.cpp (hunk 2 above)
Comment (k-ye) [reply taichi/struct/snode_tree.cpp:9]: It's mostly Google C++ Coding Style -- it suggests that mutating params being specified in the end and passed in as pointers

**Pull request merged**